R/frailty_score.R

```diff
@@ -1,152 +1,174 @@
-                             "1.0", "1.0", "1.0", "0.9", "0.9", "0.9", "0.9", "0.9", "0.9",
-                             "0.8", "0.8", "0.8", "0.8", "0.8", "0.8", "0.7", "0.7", "0.7",
-                             "0.7", "0.7", "0.6", "0.6", "0.5", "0.5", "0.5", "0.5", "0.4",
-                             "0.4", "0.4", "0.4", "0.3", "0.3", "0.1")))
-
-  frailty_score <- data.frame(frailty_score)
-
-  ### Assign scores by merging first 3 characters
-  ipdiag[, diagcode := stringr::str_sub(diagcode, 1, 3)]
-  eddiag[, diagcode := stringr::str_sub(diagcode, 1, 3)]
-  alldiag <- rbind(ipdiag, eddiag)
-
-  ipdiag <- merge(ipdiag, frailty_score, by = "diagcode", all.x = TRUE) %>%
-    .[, .(diagcode, id, score)] %>%
-    unique %>%
-    .[, .(stat = sum(as.numeric(score), na.rm = TRUE)), id]
-
-  eddiag <- merge(eddiag, frailty_score, by = "diagcode", all.x = TRUE) %>%
-    .[, .(diagcode, id, score)] %>%
-    unique %>%
-    .[, .(stat = sum(as.numeric(score), na.rm = TRUE)), id]
-
-  alldiag <- merge(alldiag, frailty_score, by = "diagcode", all.x = TRUE) %>%
-    .[, .(diagcode, id, score)] %>%
-    unique %>%
-    .[, .(stat = sum(as.numeric(score), na.rm = TRUE)), id]
-
-
-  ### merge with admission list and NA rules
-  res <- merge(res, ipdiag[, .(id, ip_frailty_score_derived = stat)], by.x = "idvar1", by.y = "id", all.x = TRUE)
-  res <- merge(res, eddiag[, .(id, er_frailty_score_derived = stat)], by.x = "idvar1", by.y = "id", all.x = TRUE)
-  res <- merge(res, alldiag[, .(id, frailty_score_derived = stat)], by.x = "idvar1", by.y = "id", all.x = TRUE)
-
-  res[(is.na(ip_frailty_score_derived) | is.na(er_frailty_score_derived)), frailty_score_derived := NA][]
-
-  data.table::setnames(res,
-                       old = c("idvar1"),
-                       new = c("genc_id"))
-  return(res)
+  ipdiag <- coerce_to_datatable(ipdiag)
+  if (!is.null(erdiag)) {erdiag <- coerce_to_datatable(erdiag)}
+  elig_enc <- unique(cohort[age >= 65, ]$genc_id)
+
+    
+  # clean and merge all diagnosis codes; return a warning if EXPLICITLY set to NULL by user
+  if (is.null(erdiag)) {
+    warning(
+      "\nBased on user input, `erdiag` is set to NULL. This is NOT recommended. The CIHI frailty score was developed and validated based on diagnosis codes from both NACRS and DAD. Excluding erdiagnosis can underestimate the level of frailty (see references in documentation for details).\n"
+      )
+    } else {
+    erdiag <- erdiag[genc_id %in% elig_enc, .(genc_id, er_diagnosis_code)] %>% rename(diagnosis_code = er_diagnosis_code)
+    }
+
+  ipdiag <- ipdiag[genc_id %in% elig_enc, .(genc_id, diagnosis_code)]
+
+  alldiag <- rbind(ipdiag, erdiag)
+
+  # Calculate score
+  frailty <- alldiag %>%
+    regex_left_join(frailty_map, by = "diagnosis_code", ignore_case = TRUE) %>% # per CIHI manual frailty conditions are identified by 595 icd-10-ca codes and any codes starting with these codes
```

Comment:
Just to avoid confusion: I believe there are actually 598 codes (see comment below: there are 3 codes missing from the mapping file), so can maybe update the comment accordingly.

Reply:
Oh interesting! Maybe it's worth mentioning this somewhere in the documentation, just so we have a record of where exactly the included codes in the final mapping file come from (i.e., CIHI document which includes additional codes that weren't mentioned in the paper)?

Reply:
Yes. The dev paper has 595 (see screenshot). The additional codes in CIHI manual will definitely need to be documented - will add to the function doc.  
![image](https://github.com/GEMINI-Medicine/Rgemini/assets/133171927/b2ac8f29-00f7-437a-a6a6-a94b45655c12)